shotover/src/unix_socket_server.rs

```diff
@@ -0,0 +1,182 @@
+
+    async fn process_request(&self, request: Request) -> Response {
+        match request {
+            Request::SendListeningSockets => {
+                info!("Processing SendListeningSockets request");
+                // TODO: In next steps, we'll extract actual file descriptors
+                let mut port_to_fd = HashMap::new();
+                port_to_fd.insert(6380, crate::hot_reload::FileDescriptor(10));
+                Response::SendListeningSockets { port_to_fd }
+            }
+        }
+    }
+}
+
+impl Drop for UnixSocketServer {
+    fn drop(&mut self) {
+        if Path::new(&self.socket_path).exists() {
+            if let Err(e) = std::fs::remove_file(&self.socket_path) {
+                warn!("Failed to remove socket file {}: {:?}", self.socket_path, e);
+            }
+        }
+    }
+}
+
+// helper function for JSON reading
+use serde::de::DeserializeOwned;
+
+async fn read_json<T: DeserializeOwned, R: AsyncReadExt + Unpin>(reader: &mut R) -> Result<T> {
+    let mut received_bytes = Vec::new();
+    loop {
+        let mut buf = [0u8; 1024];
+        let n = reader
+            .read(&mut buf)
+            .await
+            .context("Failed to read from stream")?;
+        if n == 0 {
+            return Err(anyhow::anyhow!(
+                "Connection closed before full JSON message received"
+            ));
+        }
+        received_bytes.extend_from_slice(&buf[..n]);
+        match serde_json::from_slice(&received_bytes) {
+            Ok(request) => return Ok(request),
+            Err(e) if e.is_eof() => continue,
+            Err(e) => return Err(e).context("Failed to parse JSON"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tokio::io::{AsyncReadExt, AsyncWriteExt};
+    use tokio::net::UnixStream;
+
+    #[tokio::test]
+    async fn test_unix_socket_server_basic() {
+        let socket_path = "/tmp/test-shotover-hotreload.sock";
+        // Clean up any existing socket
+        let _ = std::fs::remove_file(socket_path);
+        let server = UnixSocketServer::new(socket_path.to_string()).unwrap();
+        // Test that socket file was created
+        assert!(Path::new(socket_path).exists());
+
+        drop(server);
+        assert!(!Path::new(socket_path).exists());
+    }
+
+    #[tokio::test]
+    async fn test_request_response() {
+        let socket_path = "/tmp/test-shotover-request-response.sock";
+        // Clean up any existing socket
+        let _ = std::fs::remove_file(socket_path);
+        let mut server = UnixSocketServer::new(socket_path.to_string()).unwrap();
+        // Start server in background
+        let server_handle = tokio::spawn(async move {
+            tokio::select! {
+                _ = server.run() => {},
+                _ = tokio::time::sleep(tokio::time::Duration::from_secs(2)) => {}
+            }
+        });
+        // Give server time to start
+        tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
```

Comment:
If the system is really slow, the system might not be ready within 100ms. This kind of thing can happen in CI sometimes.

Additionally, sleeping in tests makes our test suite take longer to run, its not a big issue for one test, but it all adds up.

So lets do something like:
```rust
let mut stream = None;
for i in 0..1000 {
   if let Ok(new_stream) = UnixStream::connect(socket_path).await {
      stream = Some(new_stream)
      break;
   }
   tokio::time::sleep(Duration::from_millis(5)).await;
}
let stream = stream.unwrap();
```

It will continue executing as soon as the server has finished starting up.
If the server is completely broken it wait slightly longer than 5 seconds before giving up.